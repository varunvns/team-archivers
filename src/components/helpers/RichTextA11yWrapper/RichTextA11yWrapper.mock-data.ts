import { RichTextProps } from '@sitecore-jss/sitecore-jss-nextjs/types/components/RichText';

const defaultData: RichTextProps = {
  field: {
    value:
      '<p>At vero eos et accusamus et iusto odio dignissimos ducimus qui <a rel="noopener noreferrer" href="https://www.google.com" target="_blank" class="overlay-source">blanditiis praesentium</a> voluptatum deleniti atque corrupti quos dolores et quas molestias excepturi sint occaecati cupiditate non provident, similique sunt in culpa qui officia deserunt mollitia animi, id est laborum et dolorum fuga.</p> <ul> <li>Et harum quidem rerum facilis <a rel="noopener noreferrer" href="https://www.google.com" target="_blank">est et expedita</a> distinctio.</li> <li>Nam libero tempore, cum soluta nobis est eligendi optio cumque.</li> </ul> <p>Nihil impedit quo minus id quod maxime placeat facere possimus, omnis voluptas assumenda est, omnis dolor repellendus. Temporibus autem quibusdam et aut officiis debitis aut rerum necessitatibus saepe eveniet ut et voluptates repudiandae sint et <a href="https://www.google.com">molestiae non recusandae</a>. Itaque earum rerum hic tenetur a sapiente delectus, ut aut reiciendis voluptatibus maiores alias consequatur aut perferendis doloribus asperiores repellat.</p>',
  },
};

export const regExTest: RichTextProps = {
  ...defaultData,
  field: {
    value:
      '<p>At vero eos et accusamus et iusto odio dignissimos ducimus qui <a rel="noopener noreferrer" href="https://www.google.com" target="_blank" class="overlay-source">blanditiis praesentium</a> voluptatum deleniti atque corrupti quos dolores et quas molestias excepturi sint occaecati cupiditate non provident, similique sunt in culpa qui officia deserunt mollitia animi, id est laborum et dolorum fuga.</p> <ul> <li>Et harum quidem rerum facilis <a rel="noopener noreferrer" href="https://www.google.com" target="_blank">est et expedita</a> distinctio.</li> <li>Nam libero tempore, cum soluta nobis est eligendi optio cumque.</li> </ul> <p>Nihil impedit quo minus id quod maxime placeat facere possimus, omnis voluptas assumenda est, omnis dolor repellendus. Temporibus autem quibusdam et aut officiis debitis aut rerum necessitatibus saepe eveniet ut et voluptates repudiandae sint et\n<p>voluptatum\n<a href="adfdf" rel="adfd" target="_blank">molestiae non recusandae</a>.</p>\n<a bb="dafa" target="_blank" crazy="crazy">foobar <span>foobar</span><><><><>>>>>><<<#*)#(*Q$)JALKJFLKJALKSFJ</a>\n<a zz="adlkfjalaldkfj" bb="lakdfjl" target="_blank"></a>\n<a data-foo="adlkfjalaldkfj" bb="lakdfjl" target="_blank">adfadfadf <span>afadfd</span></a>\n<a dababd="zzzz" target="_blank"></a>\n<a dababd="zzzz"></a>\n<a 11111="zzzz" target="_blank"></a>\n<a #%%#)(*)(%*="zzz" target="_blank"><div>foobarbaz</div></a>\n<a #%%#)(*)(%*="zzz" target="_blank">\n<div>foobarbaz</div>\n</a><a #%%#)(*)(%*="zzz">\n<div>foobarbaz</div>\n</a><a #%%#)(*)(%*="zzz" target="_blank">\n<div>foobarbaz</div>\n</a>\nItaque earum rerum hic tenetur a sapiente delectus, ut aut reiciendis voluptatibus maiores alias consequatur aut perferendis doloribus asperiores repellat.</p>-',
  },
};

export default defaultData;
