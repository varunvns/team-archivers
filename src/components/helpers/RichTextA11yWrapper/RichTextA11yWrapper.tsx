// Global
import React, { useEffect, useState } from 'react';
import { Field, RichText } from '@sitecore-jss/sitecore-jss-nextjs';
import { RichTextProps } from '@sitecore-jss/sitecore-jss-nextjs/types/components/RichText';

//Local
import useExperienceEditor from 'lib/use-experience-editor';

const srOnlySpan = '<span class="sr-only"> (Opens in a new tab)</span>';

const RichTextA11yWrapper = ({ field, editable = true, ...props }: RichTextProps): JSX.Element => {
  /**
   * Hook for experience editor
   */
  const isExperienceEditor = useExperienceEditor();

  const [content, setContent] = useState(field?.value);

  const updatedField: Field<string> = {
    value: content || '',
  };

  // Run this client-side because we don't have access to the document server-side
  useEffect(() => {
    const template = document.createElement('template');
    template.innerHTML = field?.value || '';

    // Find all links either either have target="_blank" or appear to be external due to starting with "http"
    const externalLinks = [...template.content.querySelectorAll('a')].filter(
      (a) =>
        a.attributes.getNamedItem('href')?.value.startsWith('http') ||
        a.attributes.getNamedItem('target')?.value === '_blank'
    );
    // Update each external link
    externalLinks.forEach((a) => {
      // Set to open in new tab
      a.setAttribute('target', '_blank');
      // Add Screen Reader text and new tab icon
      // eslint-disable-next-line no-param-reassign
      a.innerHTML = `${a.innerHTML}${srOnlySpan}`;
    });

    // Update the content
    setContent(template.innerHTML);
  }, [field?.value]);

  // If there's no text, the only render it if it is editable and we are editing.
  // Otherwise don't render anything
  if (!field?.value) {
    if (isExperienceEditor && editable) {
      return <RichText field={field} {...props} editable={editable} className="rte" />;
    }
    return <></>;
  }

  // Just a pass through if it's EE.
  if (isExperienceEditor) {
    return <RichText field={field} {...props} editable={editable} className="rte" />;
  }
  /**
   * If not EE, use the updated field value
   */
  return <RichText field={updatedField} {...props} editable={editable} className="rte" />;
};

export default RichTextA11yWrapper;
