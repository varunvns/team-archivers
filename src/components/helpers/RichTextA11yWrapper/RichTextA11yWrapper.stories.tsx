// Global
import type { Meta, StoryObj } from '@storybook/react';
import { RichTextProps } from '@sitecore-jss/sitecore-jss-nextjs/types/components/RichText';

// Lib
import { expandObj, flattenObj } from 'lib/object-parser';

// Local
import RichTextA11yWrapper from './RichTextA11yWrapper';
import defaultData from './RichTextA11yWrapper.mock-data';

const meta: Meta<typeof RichTextA11yWrapper> = {
  title: 'Helpers/A11y/RichTextA11yWrapper',
  component: RichTextA11yWrapper,
  argTypes: {
    editable: { control: 'boolean' },
  },
};

export default meta;

type Story = StoryObj<typeof RichTextA11yWrapper>;

export const Default: Story = {
  render: (args) => {
    return <RichTextA11yWrapper {...(expandObj({ ...args }) as RichTextProps)} />;
  },
  args: {
    ...flattenObj(defaultData),
  },
};
