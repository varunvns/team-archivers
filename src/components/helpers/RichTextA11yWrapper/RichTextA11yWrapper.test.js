// Global
import { render } from '@testing-library/react';

// Lib
import { snapshot } from 'lib/testing/utils';

// Local
import RichTextA11yWrapper from './RichTextA11yWrapper';
import defaultData, { regExTest } from './RichTextA11yWrapper.mock-data';

it('renders common usage correctly', () => {
  snapshot(RichTextA11yWrapper, { componentProps: defaultData });
});

it('handles various attributes, invalid attributes, and character types in <a> tags appropriately and nested tags in <a> tags', () => {
  snapshot(RichTextA11yWrapper, { componentProps: regExTest });
});
