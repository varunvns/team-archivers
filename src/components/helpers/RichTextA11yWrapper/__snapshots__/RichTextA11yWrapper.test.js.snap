// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handles various attributes, invalid attributes, and character types in <a> tags appropriately and nested tags in <a> tags 1`] = `
<div
  class="rte"
>
  <p>
    At vero eos et accusamus et iusto odio dignissimos ducimus qui 
    <a
      class="overlay-source"
      href="https://www.google.com"
      rel="noopener noreferrer"
      target="_blank"
    >
      blanditiis praesentium
      <span
        class="sr-only"
      >
         (Opens in a new tab)
      </span>
    </a>
     voluptatum deleniti atque corrupti quos dolores et quas molestias excepturi sint occaecati cupiditate non provident, similique sunt in culpa qui officia deserunt mollitia animi, id est laborum et dolorum fuga.
  </p>
   
  <ul>
     
    <li>
      Et harum quidem rerum facilis 
      <a
        href="https://www.google.com"
        rel="noopener noreferrer"
        target="_blank"
      >
        est et expedita
        <span
          class="sr-only"
        >
           (Opens in a new tab)
        </span>
      </a>
       distinctio.
    </li>
     
    <li>
      Nam libero tempore, cum soluta nobis est eligendi optio cumque.
    </li>
     
  </ul>
   
  <p>
    Nihil impedit quo minus id quod maxime placeat facere possimus, omnis voluptas assumenda est, omnis dolor repellendus. Temporibus autem quibusdam et aut officiis debitis aut rerum necessitatibus saepe eveniet ut et voluptates repudiandae sint et

  </p>
  <p>
    voluptatum

    <a
      href="adfdf"
      rel="adfd"
      target="_blank"
    >
      molestiae non recusandae
      <span
        class="sr-only"
      >
         (Opens in a new tab)
      </span>
    </a>
    .
  </p>
  

  <a
    bb="dafa"
    crazy="crazy"
    target="_blank"
  >
    foobar 
    <span>
      foobar
    </span>
    &lt;&gt;&lt;&gt;&lt;&gt;&lt;&gt;&gt;&gt;&gt;&gt;&gt;&lt;&lt;&lt;#*)#(*Q$)JALKJFLKJALKSFJ
    <span
      class="sr-only"
    >
       (Opens in a new tab)
    </span>
  </a>
  

  <a
    bb="lakdfjl"
    target="_blank"
    zz="adlkfjalaldkfj"
  >
    <span
      class="sr-only"
    >
       (Opens in a new tab)
    </span>
  </a>
  

  <a
    bb="lakdfjl"
    data-foo="adlkfjalaldkfj"
    target="_blank"
  >
    adfadfadf 
    <span>
      afadfd
    </span>
    <span
      class="sr-only"
    >
       (Opens in a new tab)
    </span>
  </a>
  

  <a
    dababd="zzzz"
    target="_blank"
  >
    <span
      class="sr-only"
    >
       (Opens in a new tab)
    </span>
  </a>
  

  <a
    dababd="zzzz"
  />
  

  <a
    11111="zzzz"
    target="_blank"
  >
    <span
      class="sr-only"
    >
       (Opens in a new tab)
    </span>
  </a>
  

  <a
    #%%#)(*)(%*="zzz"
    target="_blank"
  >
    <div>
      foobarbaz
    </div>
    <span
      class="sr-only"
    >
       (Opens in a new tab)
    </span>
  </a>
  

  <a
    #%%#)(*)(%*="zzz"
    target="_blank"
  >
    

    <div>
      foobarbaz
    </div>
    

    <span
      class="sr-only"
    >
       (Opens in a new tab)
    </span>
  </a>
  <a
    #%%#)(*)(%*="zzz"
  >
    

    <div>
      foobarbaz
    </div>
    

  </a>
  <a
    #%%#)(*)(%*="zzz"
    target="_blank"
  >
    

    <div>
      foobarbaz
    </div>
    

    <span
      class="sr-only"
    >
       (Opens in a new tab)
    </span>
  </a>
  
Itaque earum rerum hic tenetur a sapiente delectus, ut aut reiciendis voluptatibus maiores alias consequatur aut perferendis doloribus asperiores repellat.
  <p />
  -
</div>
`;

exports[`renders common usage correctly 1`] = `
<div
  class="rte"
>
  <p>
    At vero eos et accusamus et iusto odio dignissimos ducimus qui 
    <a
      class="overlay-source"
      href="https://www.google.com"
      rel="noopener noreferrer"
      target="_blank"
    >
      blanditiis praesentium
      <span
        class="sr-only"
      >
         (Opens in a new tab)
      </span>
    </a>
     voluptatum deleniti atque corrupti quos dolores et quas molestias excepturi sint occaecati cupiditate non provident, similique sunt in culpa qui officia deserunt mollitia animi, id est laborum et dolorum fuga.
  </p>
   
  <ul>
     
    <li>
      Et harum quidem rerum facilis 
      <a
        href="https://www.google.com"
        rel="noopener noreferrer"
        target="_blank"
      >
        est et expedita
        <span
          class="sr-only"
        >
           (Opens in a new tab)
        </span>
      </a>
       distinctio.
    </li>
     
    <li>
      Nam libero tempore, cum soluta nobis est eligendi optio cumque.
    </li>
     
  </ul>
   
  <p>
    Nihil impedit quo minus id quod maxime placeat facere possimus, omnis voluptas assumenda est, omnis dolor repellendus. Temporibus autem quibusdam et aut officiis debitis aut rerum necessitatibus saepe eveniet ut et voluptates repudiandae sint et 
    <a
      href="https://www.google.com"
      target="_blank"
    >
      molestiae non recusandae
      <span
        class="sr-only"
      >
         (Opens in a new tab)
      </span>
    </a>
    . Itaque earum rerum hic tenetur a sapiente delectus, ut aut reiciendis voluptatibus maiores alias consequatur aut perferendis doloribus asperiores repellat.
  </p>
</div>
`;
